File: clients/dart/cldf/lib/models/location.dart
import 'package:json_annotation/json_annotation.dart';

import 'enums/rock_type.dart';
import 'enums/terrain_type.dart';
import 'media/media.dart';

part 'location.g.dart';

/// A climbing location (crag, gym, etc.)
@JsonSerializable(includeIfNull: false)
class Location {
  /// Creates a new [Location] instance
  Location({
    required this.id,
    this.clid,
    required this.name,
    required this.isIndoor,
    this.country,
    this.state,
    this.city,
    this.address,
    this.coordinates,
    this.starred = false,
    this.rockType,
    this.terrainType,
    this.accessInfo,
    this.media,
    this.createdAt,
    this.tags,
    this.customFields,
  });

  /// Creates a [Location] from JSON
  factory Location.fromJson(Map<String, dynamic> json) =>
      _$LocationFromJson(json);

  /// Unique identifier
  final int id;

  /// CrushLog ID - globally unique identifier (v1.3.0+)
  final String? clid;

  /// Location name
  final String name;

  /// Whether this is an indoor location
  final bool isIndoor;

  /// Country
  final String? country;

  /// State or province
  final String? state;

  /// City
  final String? city;

  /// Full address
  final String? address;

  /// GPS coordinates
  final Coordinates? coordinates;

  /// Whether this location is marked as favorite
  final bool starred;

  /// Type of rock
  final RockType? rockType;

  /// Type of terrain
  final TerrainType? terrainType;

  /// Access information
  final String? accessInfo;

  /// Media associated with this location (overview, approach photos, etc.)
  final Media? media;

  /// Creation timestamp
  final DateTime? createdAt;

  /// Associated tags
  final List<String>? tags;

  /// Custom fields
  final Map<String, dynamic>? customFields;

  /// Converts this [Location] to JSON
  Map<String, dynamic> toJson() => _$LocationToJson(this);
}

/// GPS coordinates
@JsonSerializable(includeIfNull: false)
class Coordinates {
  /// Creates a new [Coordinates] instance
  Coordinates({
    required this.latitude,
    required this.longitude,
    this.elevation,
  });

  /// Creates a [Coordinates] from JSON
  factory Coordinates.fromJson(Map<String, dynamic> json) =>
      _$CoordinatesFromJson(json);

  /// Latitude coordinate
  final double latitude;

  /// Longitude coordinate
  final double longitude;

  /// Elevation in meters
  final double? elevation;

  /// Converts this [Coordinates] to JSON
  Map<String, dynamic> toJson() => _$CoordinatesToJson(this);
}

File: clients/java/cldf-java/src/main/java/app/crushlog/cldf/models/Location.java
package app.crushlog.cldf.models;

import java.time.OffsetDateTime;
import java.util.Map;

import app.crushlog.cldf.models.enums.RockType;
import app.crushlog.cldf.models.enums.TerrainType;
import app.crushlog.cldf.models.media.Media;
import app.crushlog.cldf.utils.FlexibleDateTimeDeserializer;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/** Represents a climbing location such as a crag, boulder area, or climbing gym. */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Location {

  @JsonProperty(required = true)
  private Integer id;

  /** CrushLog ID - globally unique identifier (v1.3.0+) */
  private String clid;

  @JsonProperty(required = true)
  private String name;

  @JsonProperty(required = true)
  private Boolean isIndoor;

  private Coordinates coordinates;

  private String country;

  private String state;

  private String city;

  private String address;

  @Builder.Default private boolean starred = false;

  private RockType rockType;

  private TerrainType terrainType;

  private String accessInfo;

  /** Media associated with this location (overview, approach photos, etc.) */
  private Media media;

  @JsonDeserialize(using = FlexibleDateTimeDeserializer.class)
  private OffsetDateTime createdAt;

  private Map<String, Object> customFields;

  /** Geographic coordinates for a location. */
  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class Coordinates {
    @JsonProperty(required = true)
    private Double latitude;

    @JsonProperty(required = true)
    private Double longitude;
  }
}
